src/evict.c

```diff
@@ -595,24 +598,18 @@ int performEvictions(void) {
                  * every DB. */
                 for (i = 0; i < server.dbnum; i++) {
                     db = server.db+i;
-                    do {
-                        int slot = 0;
-                        if (server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS) {
-                            slot = getFairRandomSlot(db, DB_MAIN);
-                            dict = db->dict[slot];
-                        } else {
-                            slot = getFairRandomSlot(db, DB_EXPIRES);
-                            dict = db->expires[slot];
-                        }
-                        if ((keys = dictSize(dict)) != 0) {
-                            evictionPoolPopulate(i, slot, dict, db, pool);
-                            total_keys += keys;
-                        }
-                    /* Since keys are distributed across smaller slot-specific dictionaries in cluster mode, we may need to
-                     * visit more than one dictionary in order to populate required number of samples into eviction pool. */
-                    } while (server.cluster_enabled && keys != 0 && server.maxmemory_policy & MAXMEMORY_FLAG_ALLKEYS &&
-                        total_keys < (unsigned long) server.maxmemory_samples
-                    );
+                    if ((keys = dbSize(db, keyType)) != 0) {
+                        total_keys += keys;
+                        unsigned long sampled_keys = 0;
+                        do {
+                            int slot = getFairRandomSlot(db, keyType);
+                            dict = (keyType == DB_MAIN ? db->dict[slot] : db->expires[slot]);
+                            if (dictSize(dict) != 0) {
+                                sampled_keys += evictionPoolPopulate(i, slot, dict, db->dict[slot], pool);
+                            }
+                        } while (keys > (unsigned long) server.maxmemory_samples*10 &&
+                                 sampled_keys < (unsigned long) server.maxmemory_samples);
```

Comment:
i don't understand the `maxmemory_samples*10` part, and either way, a comment would be suitable here.

Reply:
let me see if i get this right...

we keep sampling as long as the number of keys sampled from the **current** dict is insufficient, **AND** the total number of keys in **all** the sampled databases (current and past) is more than 10 times the number of required samples.

or in other words, we exit the loop if either we got enough samples from the **current dict**, or the total number of keys in all the DBs sampled so far doesn't have enough keys in it.

i'm having a hard time following the logic here.
maybe it'll be easier to read if one of the exit conditions is an `if () break;` in the middle of the loop?

also, the fact that one variable is just from **currently** sampled **dict**, and the other is a **total** number of keys of **all** sampled **DBs** seems odd to me, and it's hard to see the symmetry here between multiple dicts in one db, and multiple dbs in one server (i'd imagine that we want both to behave similarly)